stake-pool/single-pool-cli/src/main.rs

```diff
@@ -0,0 +1,710 @@
+
+    let stake_account = Keypair::new();
+    let stake_account_address = stake_account.pubkey();
+
+    // since we cant infer pool from token account, the withdraw invocation is rather simpler
+    // first get the pool address
+    let pool_address = pool_address_from_args(
+        command_config.pool_address,
+        command_config.vote_account_address,
+    );
+
+    if config
+        .program_client
+        .get_account(pool_address)
+        .await?
+        .is_none()
+    {
+        return Err(format!("Pool {} has not been initialized", pool_address).into());
+    }
+
+    // now all the mint and token info
+    let pool_mint_address = find_pool_mint_address(&single_pool::id(), &pool_address);
+    let token = Token::new(
+        config.program_client.clone(),
+        &spl_token::id(),
+        &pool_mint_address,
+        None,
+        payer.clone(),
+    );
+
+    let token_account_address = command_config
+        .token_account_address
+        .unwrap_or_else(|| token.get_associated_token_address(&owner.pubkey()));
+
+    let token_account = token.get_account_info(&token_account_address).await?;
+
+    let token_amount = if let Some(amount) = command_config.token_amount {
+        amount
+    } else if command_config.full_withdraw {
+        token_account.base.amount
+    } else {
+        unreachable!()
+    };
+
+    println_display(
+        config,
+        format!(
+            "Withdrawing from pool {} into new stake account {}; burning {} tokens from {}\n",
+            pool_address, stake_account_address, token_amount, token_account_address,
+        ),
+    );
+
+    if token_amount == 0 {
+        return Err("Cannot withdraw zero tokens".into());
+    }
+
+    if token_amount > token_account.base.amount {
+        return Err(format!(
+            "Withdraw amount {} exceeds tokens in account ({})",
+            token_amount, token_account.base.amount
+        )
+        .into());
+    }
+
+    // note a delegate authority is not allowed here because we must authorize the pool authority
+    if token_account.base.owner != token_authority.pubkey() {
+        return Err(format!(
+            "Invalid token authority: got {}, actual {}",
+            token_account.base.owner,
+            token_authority.pubkey()
+        )
+        .into());
+    }
+
+    // looking good
+
+    let mut instructions = vec![
+        quarantine::blank_stake_instruction(config, &payer.pubkey(), &stake_account_address)
+            .await?,
+    ];
+    instructions.extend(single_pool::instruction::withdraw(
+        &single_pool::id(),
+        &pool_address,
+        &stake_account_address,
+        &stake_authority_address,
+        &token_account_address,
+        &token_authority.pubkey(),
+        token_amount,
+    ));
+
+    let transaction = Transaction::new_signed_with_payer(
+        &instructions,
+        Some(&payer.pubkey()),
+        &vec![payer, token_authority, Arc::new(stake_account)], // TODO dedupe
```

Comment:
nit: Is it possible to avoid the `Arc` and pass everything with `as_ref()` or `&` ie
```suggestion
        &vec![&payer, &token_authority, &stake_account], // TODO dedupe
```

Reply:
best i can do is `&vec![&*payer, &*token_authority, &stake_account]` but im not sure thats much better

Reply:
How about `&[payer.as_ref(), token_authority.as_ref(), &stake_account]`? If that doesn't work, we can keep yours